src/shmem_node_util.c

```diff
@@ -0,0 +1,92 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "shmem_internal.h"
+#include "shmem_node_util.h"
+
+
+char *shmem_internal_location_array = NULL;
+
+static int node_util_is_initialized = 0;
+
+
+int shmem_node_util_init(void) {
+    int ret;
+
+    if (!node_util_is_initialized) {
+        shmem_internal_location_array = malloc(sizeof(char) * shmem_internal_num_pes);
+        if (NULL == shmem_internal_location_array) {
+            RAISE_ERROR_STR("Error: out of memory when allocating node_util location array");
+            return 1;
+        }
+
+        memset(shmem_internal_location_array, -1, shmem_internal_num_pes);
+
+        ret = shmem_runtime_put("nodename", shmem_internal_nodename(), strlen(shmem_internal_nodename())+1);
+        if (ret !=0) {
+            RAISE_ERROR_MSG("shmem_node_util_init failed during nodename store to KVS: (%d)", ret);
+            return ret;
+        }
+
+        node_util_is_initialized = 1;
+    } else {
+        RAISE_WARN_STR("Attemping to initialize shmem_node_util more than once");
+    }
+
+    return 0;
+}
+
+
+inline
+void shmem_node_util_set_node_rank(int pe, int node_rank) {
+
+    shmem_internal_location_array[pe] = node_rank;
```

Comment:
Is this implicitly converting `node_rank` from `int` to `char`?  Seems very bad.

Reply:
Yikes, that does seem bad.  This was inherited from the [macro above](https://github.com/Sandia-OpenSHMEM/SOS/pull/653/files#diff-599d81d6db092bca8aafa073bfd4f93bL32).

Reply:
Yeah, this seems like the source of the 255 constant, although 255 is wrong since we set entries to -1 (also 255) for PE's not on the same node.  It would be much safer for `shmem_internal_location_array` to have the same type (`int`) as a regular SHMEM PE.  Or I suspect this will bite us in the future.